File: NHAY/NHAY_Z.py
import sys


def z_function(string, z, lenth):

    left = 0
    right = 0

    for i in xrange(1, lenth):
        if i <= right:
             z[i] = min(z[i - left], right - i + 1)
        while i + z[i] < lenth and string[z[i]] == string[i + z[i]]:
            z[i] += 1
        if i + z[i] - 1 > right:
            right = i + z[i] - 1
            left = i


def main():
    flag = 0

    for line in sys.stdin:
        if flag == 0:
            n = int(line.strip())
            flag = 1
            continue
        elif flag == 1:
            needle = line.strip()
            flag = 2
            continue
        else:
            haystack = line.strip()
            flag = 0

        string = needle + '&' + haystack
        string = "BABB"

        lenth = len(string)
        z = [0] * lenth
        z_function(string, z, lenth)
        s = 0
        print z
        for i in xrange(n + 1, lenth - n + 1):
            if z[i] == n:
                s = 1
                print i - n - 1
        if s == 1:
            sys.stdout.write("\n\n")

if __name__ == "__main__":
    main()

File: NHAY/NHAY_Z.cpp
#include<iostream>
#include<stdio.h>
#include<string>
#include<vector>
#define MAX_N 5000000
using namespace std;
string needle	;
string haystack	;
string concat   ;

long long int min(long long int a, long long int b){
	if (a > b)
		return b	;
	return a		;
}
void z_function(string s, long long int n, long long int t){
    long long int z[n]	;
    for(int i = 0; i < n;  i++)
    	z[i] = 0	;
	int s1 ;
    for(long long int i = 1, l = 0,  r = 0; i < n; i++){
        if (i <= r)
            z[i] = min(z[i - l], r - i + 1)  ;
        while(i + z[i] < n && concat[z[i]] == concat[i + z[i]])
            z[i]++ ;
        if (i + z[i] -1 > r){
            l = i   ;
            r = i + z[i] - 1 ;
        }
    }
    s1 = 0	;
    for(long long int i = t + 1; i < concat.length() - t + 1; i++){
                if (z[i] == t){
                    s1 = 1;
                    cout << i - t - 1 << "\n"  ;
            }
        }
    if (s1 == 1)
        cout << "\n\n"  ;
}
int main()  {
    long long int t ;
    while(scanf ("%lld\n",  &t) == 1){
        	cin >> needle >> haystack	;
        	concat = needle + '&' + haystack  ;
        	z_function(concat,concat.length(),t)    ;
    }
}
